File: 4. Tribonacci Sequence.cpp
// 4. Tribonacci Sequence
// Topic: Tribonacci
// Description: Compute nth Tribonacci number using recursion + memoization.
// Input: n (integer >= 0)
// Output: nth Tribonacci number

#include <bits/stdc++.h>
using namespace std;

// Memoization starts here: vector<long long> memo
long long trib(int n, vector<long long>& memo)
{
    if (n == 0)
        return 0;
    if (n == 1 || n == 2)
        return 1;
    if (memo[n] != -1)
        return memo[n];
    memo[n] = trib(n - 1, memo) + trib(n - 2, memo) + trib(n - 3, memo);
    return memo[n];
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<long long> memo(n + 1, -1);
    cout << trib(n, memo) << '\n';
}

// Memoization process started in trib function where memo vector is accessed.
// Time Complexity: O(n)
// Space Complexity: O(n)

File: 4. Tribonacci Sequence.py
# 4. Tribonacci Sequence
# Topic: Tribonacci Sequence
# Description: Each term is the sum of the previous 3 terms; compute the nth Tribonacci number using recursion + memoization.
# Input: n (integer >= 0)
# Output: nth Tribonacci number

import sys
from functools import lru_cache

@lru_cache(maxsize=None)
def trib(n: int) -> int:
    if n == 0:
        return 0
    if n == 1 or n == 2:
        return 1
    return trib(n-1) + trib(n-2) + trib(n-3)

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    n = int(data[0])
    print(trib(n))

# Memoization process started at trib decorated with lru_cache.
# Time Complexity: O(n)
# Space Complexity: O(n)